### PR TITLE
Time evolution implemented for Synchrotron Self Compton

### PR DESCRIPTION
**Pull request by grzegorzbor**
SSC time evolution implemented

### CODE REVIEW

File: agnpy/compton/synchrotron_self_compton.py [hunk 1]
```diff
@@ -33,6 +33,36 @@ def __init__(self, blob, ssa=False, integrator=np.trapz):
...
+            self.blob.R_b,
+            self.blob.n_e,
+            *self.blob.n_e.parameters,
+            gamma=gamma
+        )
+
+        # differential energy density
+        u_epsilon0 = (3 * np.power(self.blob.d_L, 2) * sed_flux_synch) / (
+                c * np.power(self.blob.R_b, 2) * np.power(self.blob.delta_D, 4) * epsilon0
+        )
```
Comment (jsitarek) [inline agnpy/compton/synchrotron_self_compton.py:57]: I think you need an extra factor of 3/4 here, compere with this:
https://github.com/cosimoNigro/agnpy/blame/506c038fa751369e63e20a85a222558a8997b8d4/agnpy/compton/synchrotron_self_compton.py#L109-L111

File: agnpy/emission_regions/blob.py [hunk 2]
```diff
@@ -393,3 +397,50 @@ def u_ph_synch(self):
             * np.trapz(np.power(self.gamma_e, 2) * self.n_e(self.gamma_e), self.gamma_e)
         )
         return u_ph.to("erg cm-3")
+
+    def n_e_time_evolution(self, energy_loss_function, time, subintervals_count=10):
+        """Performs the time evolution of the electron distribution inside the blob.
+
+        Parameters
+        ----------
+        energy_loss_function : function
```
Comment (jsitarek) [inline agnpy/emission_regions/blob.py:406]: this part you might need to rethink. You specify here a single function, while in reality there will be different processes to be considered and they will have different effect at different moments and at varying energies. So either specify here a list of functions (I guess it is possible??), and loop over them, or a user will need to write a function where he adds a few processes and specifies this function here.

File: agnpy/emission_regions/blob.py [hunk 3]
```diff
@@ -393,3 +397,50 @@ def u_ph_synch(self):
...
+            raise ValueError("subintervals_count must be > 0")
+
+        unit_time_interval = time / subintervals_count
+
+        def gamma_recalculated_after_loss(gamma):
+            old_energy = (gamma * mec2).to("erg")
+            energy_loss_per_time = energy_loss_function(gamma).to("erg s-1")
+            energy_loss = (energy_loss_per_time * unit_time_interval).to("erg")
+            new_energy = old_energy - energy_loss
+            if np.any(new_energy < 0):
```
Comment (jsitarek) [inline agnpy/emission_regions/blob.py:423]: this we discussed in person. If the new energy is negative then the code has failed miserably with way too large time step. However if the loss is e.g. 50% is also is very large, because energy losses at E/2 will be different. And limiting the global step size can be inefficient because at low energies nothing is happening. 
The most efficient solution might be try a given step, but at energies where it is too large (e.g. change of energy is more then 10%) the step is divided iteratively by 2 until the change is small enough at this energy.

File: agnpy/spectra/tests/test_spectra.py [hunk 4]
```diff
@@ -760,12 +760,13 @@ def test_compare_numerical_results_with_analytical_calculation(self, p, time_and
         initial_n_e = PowerLaw(k, p, gamma_min=gamma_min, gamma_max=gamma_max, mass=m_e)
         blob = Blob(1e16 * u.cm, Distance(1e27, unit=u.cm).z, 0, 10, 1 * u.G, n_e=initial_n_e)
         synch = Synchrotron(blob)
-        evaluated_n_e = initial_n_e.evaluate_time(time, synch.electron_energy_loss_per_time, subintervals_count=steps)
+        blob.n_e_time_evolution(synch.electron_energy_loss_rate, time, subintervals_count=steps)
+        evaluated_n_e = blob.n_e
```
Comment (jsitarek) [inline agnpy/spectra/tests/test_spectra.py:764]: be careful with this, because I suspect that this will make a shallow copy of the n_e

File: agnpy/spectra/tests/test_spectra.py [hunk 5]
```diff
@@ -803,11 +804,14 @@ def test_compare_calculation_with_calculation_split_into_two(self):
         synch = Synchrotron(blob)
 
         # iterate over 60 s in 20 steps
-        eval_1 = initial_n_e.evaluate_time(60*u.s, synch.electron_energy_loss_per_time, subintervals_count=20)
+        blob.n_e_time_evolution(synch.electron_energy_loss_rate, 60 * u.s, subintervals_count=20)
+        eval_1 = blob.n_e
```
Comment (jsitarek) [inline agnpy/spectra/tests/test_spectra.py:808]: another shallow copy in eval_1 and eval_2

File: agnpy/spectra/tests/test_spectra.py [hunk 6]
```diff
@@ -817,4 +821,62 @@ def test_compare_calculation_with_calculation_split_into_two(self):
...
+        )
+
+        blob = Blob(R_b, Distance(1e27, unit=u.cm).z, 10, 10, 1 * u.G, n_e=n_e)
+        synch = Synchrotron(blob)
+        ssc = SynchrotronSelfCompton(blob)
+
+        time = 60 * u.s
+        intervals = 600
+
+        eletronsBefore = n_e.integrate(n_e.gamma_min, n_e.gamma_max)
```
Comment (jsitarek) [inline agnpy/spectra/tests/test_spectra.py:842]: typo in the variable name: electronsBefore

File: agnpy/spectra/tests/test_spectra.py [hunk 7]
```diff
@@ -817,4 +821,62 @@ def test_compare_calculation_with_calculation_split_into_two(self):
...
+        )
+
+        blob = Blob(R_b, Distance(1e27, unit=u.cm).z, 10, 10, 1 * u.G, n_e=n_e)
+        ssc = SynchrotronSelfCompton(blob)
+
+        time = 60 * u.s
+        intervals = 200
+
+        blob.n_e_time_evolution(ssc.electron_energy_loss_rate_thomson, time, subintervals_count=intervals)
+        n_e_thompson = blob.n_e
```
Comment (jsitarek) [inline agnpy/spectra/tests/test_spectra.py:877]: possibly a shallow copy here

File: agnpy/synchrotron/synchrotron.py [hunk 8]
```diff
@@ -90,18 +90,13 @@ def __init__(self, blob, ssa=False, integrator=np.trapz):
...
-        fixed = self._electron_energy_loss_formula_prefix()
+    def electron_energy_loss_rate(self, gamma):
+        fixed = self._electron_energy_loss_formula_prefactor()
         return fixed * gamma ** 2
 
-    def _electron_energy_loss_formula_prefix(self):
-        # using SI units here because of the ambiguity of the different CGS systems in terms of mu0 definition
-        magn_energy_dens = (self.blob.B.to("T") ** 2) / (2 * mu0)
-        return ((4 / 3) * sigma_T * c * magn_energy_dens).to("erg/s")
+    def _electron_energy_loss_formula_prefactor(self):
```
Comment (jsitarek) [inline agnpy/synchrotron/synchrotron.py:97]: not sure if it makes sense to split it into two functions, especially that you do not do it for IC losses

Comment (grzegorzbor) [reply agnpy/synchrotron/synchrotron.py:97]: It's split, because the "prefactor"  calculation is done in some test, so I didn't want to duplicate the formulate, and instead extracted it to the (sub-)function for reuse in those tests

File: agnpy/emission_regions/blob.py (hunk 2 above)
Comment (grzegorzbor) [reply agnpy/emission_regions/blob.py:406]: Yes I agree, this will need to be changed, but for now it's easier to use it and test it this way, and it's also quite generic. But when we add support for more processes, this part will need some rethinking, how to combine such multiple processes.

File: agnpy/spectra/tests/test_spectra.py (hunk 4 above)
Comment (grzegorzbor) [reply agnpy/spectra/tests/test_spectra.py:765]: As we already discussed, in this case it's safe because i replace the n_e in blob at each calculation step with a new copy. But just in case the internal implementation changes in the future, i added deep copy here

File: agnpy/compton/synchrotron_self_compton.py [hunk 9]
```diff
@@ -33,6 +33,40 @@ def __init__(self, blob, ssa=False, integrator=np.trapz):
...
+        # factor 3 / 4 for averaging in a sphere - see comment inside the evaluate_sed_flux method for u_synch calculation
+        u_epsilon0 *= 3 / 4
+
+        # formulas from paragraph 2.1, Moderski2005
+        b_ = 4 * gamma.reshape(len(gamma),1) * epsilon0.value.reshape(1,len(epsilon0)) # matrix [gamma x epsilon]
+        b_flattened = b_.reshape(b_.shape[0] * b_.shape[1])
+        f_KN_flattened = np.array([1 / (1 + b) ** 1.5 if b < 1e4 else (9 / (2 * b**2)) * (log(b) - 11/6) for b in b_flattened])
+        f_KN = f_KN_flattened.reshape(len(gamma), len(epsilon0)) # back to [gamma x epsilon] matrix
```
Comment (jsitarek) [inline agnpy/compton/synchrotron_self_compton.py:65]: I think this double reshape and for loop also adds to the execution time of the code. You can do it simpler and probably faster using np.where and operating directly on the matrix

File: agnpy/spectra/tests/test_spectra.py [hunk 10]
```diff
@@ -803,11 +810,14 @@ def test_compare_calculation_with_calculation_split_into_two(self):
...
+        blob.n_e_time_evolution(synch.electron_energy_loss_rate, 60 * u.s, subintervals_count=20)
+        eval_1 = deepcopy(blob.n_e)
         # iterate first over 30 s, and then, starting with interpolated distribution, over the remaining 30 s,
         # with slightly different number of subintervals
-        eval_2 = initial_n_e.evaluate_time(30 * u.s, synch.electron_energy_loss_per_time, subintervals_count=10)\
-            .evaluate_time(30 * u.s, synch.electron_energy_loss_per_time, subintervals_count=8)
+        blob.n_e = initial_n_e
+        blob.n_e_time_evolution(synch.electron_energy_loss_rate, 30 * u.s, subintervals_count=10)
+        blob.n_e_time_evolution(synch.electron_energy_loss_rate, 30 * u.s, subintervals_count=8)
+        eval_2 = deepcopy(blob.n_e)
```
Comment (jsitarek) [inline agnpy/spectra/tests/test_spectra.py:820]: here in principle you do not need deepcopy, because eval_2 is not being modified anymore

File: agnpy/spectra/tests/test_spectra.py [hunk 11]
```diff
@@ -817,4 +827,63 @@ def test_compare_calculation_with_calculation_split_into_two(self):
...
+        n_e = PowerLaw.from_total_energy(
+            1e48 * u.erg,
+            4 / 3 * np.pi * R_b ** 3,
+            p=2.8,
+            gamma_min=1e2,
+            gamma_max=1e7,
+            mass=m_e,
+        )
+
+        blob = Blob(R_b, Distance(1e27, unit=u.cm).z, 10, 10, 1 * u.G, n_e=n_e)
```
Comment (jsitarek) [inline agnpy/spectra/tests/test_spectra.py:841]: this blob is being used in a bunch of tests, maybe make it a fixture?

File: agnpy/spectra/tests/test_spectra.py [hunk 12]
```diff
@@ -817,4 +827,63 @@ def test_compare_calculation_with_calculation_split_into_two(self):
...
+
+        time = 60 * u.s
+        intervals = 600
+
+        electronsBefore = n_e.integrate(n_e.gamma_min, n_e.gamma_max)
+
+        blob.n_e_time_evolution(synch.electron_energy_loss_rate, time, subintervals_count=intervals)
+        electronsAfterSync = blob.n_e.integrate(blob.n_e.gamma_min, blob.n_e.gamma_max)
+
+        blob.n_e = n_e
```
Comment (jsitarek) [inline agnpy/spectra/tests/test_spectra.py:853]: I am wondering if it is enough to only overwrite n_e, or do you need to restore the original gamma factors as well, because those are also modified in the time evolution function

File: agnpy/spectra/tests/test_spectra.py [hunk 13]
```diff
@@ -817,4 +827,63 @@ def test_compare_calculation_with_calculation_split_into_two(self):
...
+        ssc = SynchrotronSelfCompton(blob)
+
+        time = 60 * u.s
+        intervals = 200
+
+        blob.n_e_time_evolution(ssc.electron_energy_loss_rate_thomson, time, subintervals_count=intervals)
+        n_e_thompson = deepcopy(blob.n_e)
+
+        blob.n_e = n_e
+        blob.n_e_time_evolution(ssc.electron_energy_loss_rate, time, subintervals_count=intervals)
```
Comment (jsitarek) [inline agnpy/spectra/tests/test_spectra.py:887]: this effectively compares that the time_evolution gives the same with two different energy_loss_rate functions that should be close in this energy range. I think it would be better to simply compare the output of electron_energy_loss_rate and ..._thomson on a range of gammas without executing the time_evolution. On the other hand, a test that we were discussing was to compute this analytically (that is relatively simply for Thomson range) and introduce this comparison here.

File: agnpy/spectra/tests/test_spectra.py (hunk 12 above)
Comment (grzegorzbor) [reply agnpy/spectra/tests/test_spectra.py:856]: fixed, i reset the gamma params in the n_e setter

File: agnpy/emission_regions/blob.py [hunk 14]
```diff
@@ -130,12 +137,24 @@ def set_delta_D(self, Gamma, theta_s):
         mu_s = np.cos(theta_s.to("rad").value)
         self.delta_D = 1 / (self.Gamma * (1 - self.Beta * mu_s))
 
+    def reset_gamma_e(self, gamma_e_size=None):
+        if gamma_e_size is not None:
+            self.gamma_e_size = gamma_e_size
+        self.gamma_e_min = self._n_e.gamma_min
+        self.gamma_e_max = self._n_e.gamma_max
+
     def set_gamma_e(self, gamma_size, gamma_min=1, gamma_max=1e8):
```
Comment (jsitarek) [inline agnpy/emission_regions/blob.py:146]: with the reset_gamma_[ep] functions introduced, are the set_gamma_[ep] functions still used anywhere in the code?

File: agnpy/spectra/tests/test_spectra.py [hunk 15]
```diff
@@ -817,4 +824,107 @@ def test_compare_calculation_with_calculation_split_into_two(self):
             eval_2.evaluate(gammas, 1, gamma_min, gamma_max),
             0.001)
 
+    def test_total_number_of_electrons_should_stay_after_time_evolution(self):
+        r_b = 1e16 * u.cm
+        n_e = PowerLaw.from_total_energy(
+            1e48 * u.erg,
+            4 / 3 * np.pi * r_b ** 3,
+            p=2.8,
```
Comment (jsitarek) [inline agnpy/spectra/tests/test_spectra.py:832]: it might be wise to use a harder spectral index (e.g. -1.5 instead of -2.8) for this test, because this will increase the number of electrons in the high edge of the spectrum that are most affected by the cooling

File: agnpy/spectra/tests/test_spectra.py [hunk 16]
```diff
@@ -817,4 +824,107 @@ def test_compare_calculation_with_calculation_split_into_two(self):
...
+
+        ssc = SynchrotronSelfCompton(blob)
+        blob.n_e_time_evolution(ssc.electron_energy_loss_rate, time, subintervals_count=steps)
+        evaluated_n_e: InterpolatedDistribution = blob.n_e
+
+        def gamma_before(gamma_after_time, before_time):
+            coef = ssc._electron_energy_loss_thomson_formula_prefactor() / (m_e * c ** 2)
+            return 1 / ((1 / gamma_after_time) - before_time * coef)
+
+        def integral_analytical(gamma_bottom, gamma_up):
```
Comment (jsitarek) [inline agnpy/spectra/tests/test_spectra.py:911]: I think this function is being used in two different test functions - it can be extracted out of the functions to be included only once (and I would change the name to intergral_powerlaw or something like this

Comment (grzegorzbor) [reply agnpy/spectra/tests/test_spectra.py]: It turned out there is already sch function in this file, so i removed mine.

File: agnpy/spectra/tests/test_spectra.py (hunk 15 above)
Comment (grzegorzbor) [reply agnpy/spectra/tests/test_spectra.py]: I just tried with 1.5, but  it made the loss rate too high, and the time had to be split into too many subintervals. So i change it to 2.0 and reduced the total time.

File: agnpy/emission_regions/blob.py (hunk 14 above)
Comment (grzegorzbor) [reply agnpy/emission_regions/blob.py]: Removed. Also, the reset_gamma_[ep] were not really needed, so i removed them too. 
I'd like to remove the gamma_[ep]_size as well, because they don't describe the blob physical parameters. But it needs more work so I'll get back to it later.

File: agnpy/emission_regions/blob.py (hunk 3 above)
Comment (grzegorzbor) [reply agnpy/emission_regions/blob.py:427]: I'll work on it in the separate branch.

File: agnpy/spectra/tests/test_spectra.py (hunk 13 above)
Comment (grzegorzbor) [reply agnpy/spectra/tests/test_spectra.py]: Fixed
